### PR TITLE
Add a driver to support Microsoft.Data.SqlClient provider

### PR DESCRIPTION
**Pull request by hazzik**
This is just a cleaned-up copy of SqlClientDriver.

### CODE REVIEW

**Comment by fredericDelaporte** [comment]
So for adding some context, this is about following Microsoft move on their data provider, as explained [here](https://devblogs.microsoft.com/dotnet/introducing-the-new-microsoftdatasqlclient/).
This data provider targets netcoreapp2.1, netfx46, netstandard2.0, while NHibernate targets netcoreapp2.0.
So there is a mismatch. It is not blocking due to using reflection. And well, since consumers are supposed to explicitly add the dependency on the data provider anyway, they should discover by themselves they have to target netcoreapp2.1 at least.
(I have not looked into this in details, so I am just commenting, not reviewing yet.)

**Comment by hazzik** [comment]
This is pity that they claim to support netstandard 2.0 while this is not true.

**Comment by bahusoid** [comment]
>  they should discover by themselves they have to target netcoreapp2.1 at least.

Or use netstandard2.0 version of the lib.

> This is pity that they claim to support netstandard2.0 while this is not true.

Why it's not true? netcoreapp2.1 might have some target specific optimizations missing in netstandard2.0

**Comment by hazzik** [comment]
@bahusoid everything in netstandard2.0 throwing PlatformNotSupportedException.

**Comment by bahusoid** [comment]
It seems going to be fixed according to https://github.com/dotnet/SqlClient/issues/204

File: src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+		///     The SqlClient driver does NOT support more than 1 open DbDataReader
+		///     with only 1 DbConnection.
+		/// </summary>
+		/// <value><see langword="false" /> - it is not supported.</value>
+		/// <remarks>
+		///     MS SQL Server 2000 (and 7) throws an exception when multiple DbDataReaders are
+		///     attempted to be opened.  When SQL Server 2005 comes out a new driver will be
+		///     created for it because SQL Server 2005 is supposed to support it.
+		/// </remarks>
+		public override bool SupportsMultipleOpenReaders => false;
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs]: Ok.

File: src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+		public override bool SupportsMultipleQueries => true;
+
+		/// <summary>
+		///     With read committed snapshot or lower, SQL Server may have not actually already committed the transaction
+		///     right after the scope disposal.
+		/// </summary>
+		public override bool HasDelayedDistributedTransactionCompletion => true;
+
+		/// <inheritdoc />
+		public override DateTime MinDate => new DateTime(1753, 1, 1);
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs]: Yeah, should have used Sql2008ClientDriver as a baseline

File: src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+				case DbType.DateTime2:
+					dbParam.Size = MsSql2000Dialect.MaxDateTime2;
+					break;
+				case DbType.DateTimeOffset:
+					dbParam.Size = MsSql2000Dialect.MaxDateTimeOffset;
+					break;
+				case DbType.Xml:
+					dbParam.Size = MsSql2005Dialect.MaxSizeForXml;
+					break;
+			}
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs]: Thanks for noting.

**Comment by fredericDelaporte** [comment]
You have resolved a bunch of the review comments, but no new commit has been pushed matching them. Is it on purpose?

**Comment by hazzik** [comment]
> but no new commit has been pushed 

I've forgot to `git push`

**Comment by hazzik** [comment]
Actually, I pushed to incorrect remote.

File: src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs]: Actually, @fredericDelaporte this option seems to be an extension to `MultipleActiveResultSets=True/False` setting on connection string (it's off by default). If MARS is supported (SQL >= 2005) and enabled then this is true, if not - then is false. Should we account for this?

[the remark this answers is not in the document]
Comment (fredericDelaporte) [reply src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs]: Then ideally, it should be left as an option to activate on user side. So well, to keep it simple for now, just leave it false I think, but update the old comment, something like : `SQL Server 2005 and above supports this, but only as an option not enabled by default. Assuming it is not enabled is safer.`

Off-topic: by the way about data provider specific options required in many dialects or drivers, maybe we should introduce something like the `cache\.[\w._]+` setting we already have (slightly fixed by the way: `_` is already in `\w`, that is `\-` which should be added), in order to cease having to declare them all in NHibernate configuration handling. Something like `db-specific\.[\w.\-]`, with as usage convention to continue with the db name then the setting name.)

**Comment by fredericDelaporte** [comment]
Checking the test suite, there is a number of trouble with this new data provider.
 - ExceptionsTest.SQLExceptionConversionTest are no more receiving the expected exception.
 - Two Linq.QueryLock tests are failing.
 - Be it when using SQL 2012 dialect or 2005, Time types tests are failing due to expecting DateTime instead of TimeSpan (that may be due the changes I suggested, but seems strange.)
 - Tools.hbm2ddl.SchemaMetadataUpdaterTest.SchemaMetadataUpdaterFixture.EnsureReservedWordsHardCodedInDialect fails, not finding datetime2 and datetimeoffset. Hum, is this data-provider actually targeting SQL 2005 instead of being for brand new features?!
 - Min date test fails too, complaining about the date being out of the old date range... (`TypesTest.DateTypeFixture.ReadWriteMin`).

File: src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+			_dialect = Dialect.Dialect.GetDialect(settings);
+		}
+
+		/// <inheritdoc />
+		protected override void InitializeParameter(DbParameter dbParam, string name, SqlType sqlType)
+		{
+			base.InitializeParameter(dbParam, name, sqlType);
+
+			// Defaults size/precision/scale
+			switch (dbParam.DbType)
```
Comment (fredericDelaporte) [inline src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs]: There is a subtle difference in our Sql2008 driver here : it reads `sqlType.DbType`, not the `dbParam` one. And due to an [old legacy](https://github.com/dotnet/SqlClient/issues/5) they have unfortunately kept in Microsoft.Data.SqlClient, this changes a lot: `DbType.Date` and `DbType.Time` are changed to `DbType.DateTime` on the parameter.
So it is critical to read `sqlType.DbType` here instead.

That is the reason being the test failures I was reporting.

File: src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+		/// <summary>
+		///     With read committed snapshot or lower, SQL Server may have not actually already committed the transaction
+		///     right after the scope disposal.
+		/// </summary>
+		public override bool HasDelayedDistributedTransactionCompletion => true;
+
+		/// <inheritdoc />
+		public override DateTime MinDate => new DateTime(1753, 1, 1);
+
+		System.Type IEmbeddedBatcherFactoryProvider.BatcherFactoryClass => typeof(GenericBatchingBatcherFactory);
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs]: I think maybe in a later PR.

File: src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs (hunk 4 above)
Comment (hazzik) [reply src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs]: Fixed

File: src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs]: > So well, to keep it simple for now, just leave it false I

Done.

File: src/NHibernate.Test/NHSpecificTest/NH2420/Fixture.cs [hunk 6]
```diff
@@ -62,10 +62,8 @@ public void ShouldBeAbleToReleaseSuppliedConnectionAfterDistributedTransaction()
 						new DummyEnlistment(),
 						EnlistmentOptions.None);
 
-					if (Sfi.ConnectionProvider.Driver.GetType() == typeof(OdbcDriver))
-						connection = new OdbcConnection(connectionString);
-					else
-						connection = new SqlConnection(connectionString);
+					connection = Sfi.ConnectionProvider.Driver.CreateConnection();
+					connection.ConnectionString = connectionString;
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/NH2420/Fixture.cs]: `Driver.CreateConnection()` should have been used from the start. It is meant for this. It just creates the connection, not doing anything else with it.

**Comment by andreyciocan** [comment]
Hello everyone. I just tried to publish the master project and include it in my project. The problem is that MicrosoftDataSqlClient is not available under Nhibernate.Driver. Also, if I try to set the connection.driver_class to MicrosoftDataSqlClient it throws me an error that the driver cannot be created. Any hints?

**Comment by hazzik** [comment]
> Any hints?

Yes: it is not yet released.

**Comment by fredericDelaporte** [comment]
You may use binaries from the CI builds if you wish. See the [commits](https://github.com/nhibernate/nhibernate-core/commits/master), click the green check then follow the Details link from "NHibernate (Release Package)" build. (Latest [link](https://teamcity.jetbrains.com/viewLog.html?buildId=2728372&buildTypeId=NHibernate_NHibernateReleasePackage)) as of this writing.) Then login with any TeamCity account or as Guest, and go to the Artifact tabs. It will contain .Net Framework binaries and NuGet packages build from master.

**Comment by andreyciocan** [comment]
> > Any hints?
> 
> Yes: it is not yet released.

Thank's. Any idea when it will be released?

**Comment by bahusoid** [comment]
> The problem is that MicrosoftDataSqlClient is not available under Nhibernate.Driver

There is no such class even on master. It's `MicrosoftDataSqlClientDriver` and not `MicrosoftDataSqlClient`

**Comment by andreyciocan** [comment]
As @fredericDelaporte suggested, I downloaded the Nuget package and installed Nhibernate from Local. I am using the following code to configure Nhibernate:
`var cfg = new Configuration();
            cfg.SetProperty("connection.connection_string", configuration["ConnectionStrings:MyConnection"]);
            cfg.SetProperty("dialect", "Project.DB.DataAccess.Dynamic.CustomDialect, Project.DB.DataAccess");
            cfg.SetProperty("connection.driver_class", "Nhibernate.Driver.MicrosoftDataSqlClientDriver, Nhibernate");
            return cfg;`

When I try to run a db select command, the following exception is thrown: 
`NHibernate.HibernateException : Could not create the driver from Nhibernate.Driver.MicrosoftDataSqlClientDriver, Nhibernate.
      ----> System.TypeLoadException : Could not load type 'Nhibernate.Driver.MicrosoftDataSqlClientDriver' from assembly 'NHibernate, Version=5.2.0.0, Culture=neutral, PublicKeyToken=aa95f207798dfdb4'.`

**Comment by bahusoid** [comment]
Common... Just type it right... How come namespace become "Nhibernate.Driver"?? 

And avoid using strings when you can use typed values. Something like:
```C#
cfg.SetProperty("connection.driver_class", typeof(MicrosoftDataSqlClientDriver).AssemblyQualifiedName);
```

**Comment by andreyciocan** [comment]
@bahusoid Thank you! It was a silly mistake...

**Comment by luisabreu** [comment]
Hello guys.

quick question: when will NH depend on Microsoft.Data.SqlClient instead of the System.Data.SqlClient? It's suposed to be merged, but it seems like NH's nuget package still references System.Data.SqlClient assembly...

**Comment by luisabreu** [comment]
Hello guys.

quick question: when will NH depend on Microsoft.Data.SqlClient instead of the System.Data.SqlClient? It's suposed to be merged, but it seems like NH's nuget package still references System.Data.SqlClient package...

**Comment by fredericDelaporte** [comment]
This merge here does not change anything to the existing SqlClient drivers. It adds a new one, named `MicrosoftDataSqlClientDriver`. So instead of configuring NHibernate to use `SqlClientDriver` or `Sql2008ClientDriver`, configure it to use `MicrosoftDataSqlClientDriver`.

As this PR is in the 5.3 milestone, this also requires you to use at least a v5.3 NHibernate, currently available only as preview in the [NHibernate MyGet feed](https://github.com/nhibernate/nhibernate-core#nightly-development-builds).

**Comment by EngSayed** [comment]
I just switched from `System.Data.SqlClient` to `Microsoft.Data.SqlClient` and I added this configuration `connection.driver_class` to use `MicrosoftDataSqlClientDriver` but I am getting the following exception inside `SqlClientBatchingBatcher`.`AddToBatch(..)`

```
Unable to cast object of type 'Microsoft.Data.SqlClient.SqlCommand' to type 'System.Data.SqlClient.SqlCommand'.
Description: An unhandled exception occurred during the execution of the current web request. Please review the stack trace for more information about the error and where it originated in the code.

Exception Details: System.InvalidCastException: Unable to cast object of type 'Microsoft.Data.SqlClient.SqlCommand' to type 'System.Data.SqlClient.SqlCommand'.
```

I have the following config:
```
[Environment.BatchStrategy] = typeof(CustomDeleteSqlBatcherFactory).AssemblyQualifiedName,
[Environment.TransactionStrategy] = typeof(AdoNetWithSystemTransactionFactory).AssemblyQualifiedName,
[Environment.Dialect] = typeof(FlintMsSql2008Dialect).AssemblyQualifiedName,
[Environment.Isolation] = "ReadCommitted",
[Environment.ReleaseConnections] = "on_close",
[Environment.ConnectionProvider] = typeof(DriverConnectionProvider).AssemblyQualifiedName,
[Environment.ConnectionDriver] = typeof(MicrosoftDataSqlClientDriver).AssemblyQualifiedName,
[Environment.ProxyFactoryFactoryClass] = typeof(StaticProxyFactoryFactory).AssemblyQualifiedName,
```

**Comment by EngSayed** [comment]
Our `CustomDeleteSqlBatcher` inherits from `SqlClientBatchingBatcher` which is causing that exception to be thrown. Is there another class to inherit from which supports `Microsoft.Data.SqlClient`?

**Comment by fredericDelaporte** [comment]
See #3258.

**Comment by EngSayed** [comment]
> See #3258. (TLDR: use `GenericBatchingBatcher`.)

Thanks @fredericDelaporte, I found about the generic batcher. I tried it and exception was gone. Could you please let me know if there any performance difference While using the generic batcher? - just wondering

**Comment by fredericDelaporte** [comment]
That is discussed in #3258 too. We are not aware of any significant performance difference, but we have not done extensive benchmarks to check this.

**Pull request merged**